larpandoracontent/LArHelpers/LArHierarchyHelper.cc

```diff
@@ -0,0 +1,1171 @@
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/PdgTable.h"
+#include "Pandora/StatusCodes.h"
+
+#include "larpandoracontent/LArHelpers/LArHierarchyHelper.h"
+
+namespace lar_content
+{
+
+using namespace pandora;
+
+LArHierarchyHelper::MCHierarchy::MCHierarchy(const ReconstructabilityCriteria &recoCriteria) :
+    m_recoCriteria(recoCriteria),
+    m_pNeutrino{nullptr}
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+LArHierarchyHelper::MCHierarchy::~MCHierarchy()
+{
+    for (const Node *pNode : m_rootNodes)
+        delete pNode;
+    m_rootNodes.clear();
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void LArHierarchyHelper::MCHierarchy::FillHierarchy(const MCParticleList &mcParticleList, const CaloHitList &caloHitList,
+    const bool foldToPrimaries, const bool foldToLeadingShowers)
+{
+    auto predicate = [](const MCParticle *pMC) { return std::abs(pMC->GetParticleId()) == NEUTRON; };
+    m_mcToHitsMap.clear();
+    for (const CaloHit *pCaloHit : caloHitList)
+    {
+        try
+        {
+            const MCParticle *pMC{MCParticleHelper::GetMainMCParticle(pCaloHit)};
+            m_mcToHitsMap[pMC].emplace_back(pCaloHit);
+        }
+        catch(const StatusCodeException&)
+        {
+        }
+    }
+
+    if (foldToPrimaries && !foldToLeadingShowers)
+    {
+        MCParticleSet primarySet;
+        m_pNeutrino = LArHierarchyHelper::GetMCPrimaries(mcParticleList, primarySet);
+        MCParticleList primaries(primarySet.begin(), primarySet.end());
+        if (m_recoCriteria.m_removeNeutrons)
+            primaries.erase(std::remove_if(primaries.begin(), primaries.end(), predicate), primaries.end());
+        for (const MCParticle *pPrimary : primaries)
+        {
+            MCParticleList allParticles{pPrimary};
+            if (!m_recoCriteria.m_removeNeutrons)
+            {
+                LArMCParticleHelper::GetAllDescendentMCParticles(pPrimary, allParticles);
+            }
+            else
+            {   // Collect track-like and shower-like particles together, but throw out neutrons and descendents
+                MCParticleList neutrons;
+                LArMCParticleHelper::GetAllDescendentMCParticles(pPrimary, allParticles, allParticles, neutrons);
+            }
+            CaloHitList allHits;
+            for (const MCParticle *pMC : allParticles)
+            {
+                // Not all MC particles will have hits
+                if (m_mcToHitsMap.find(pMC) != m_mcToHitsMap.end())
+                {
+                    const CaloHitList &caloHits(m_mcToHitsMap.at(pMC));
+                    allHits.insert(allHits.begin(), caloHits.begin(), caloHits.end());
+                }
+            }
+            m_rootNodes.emplace_back(new Node(*this, allParticles, allHits));
```

Comment:
Worth running this through valgrind, given heap allocations. (Quick and easy to do, these days.)